addon/-private/system/relationships/state/has-many.js

```diff
@@ -111,6 +111,32 @@ export default class ManyRelationship extends Relationship {
     super.removeCanonicalInternalModelFromOwn(internalModel, idx);
   }
 
+  unloadInverseInternalModelFromOwn(internalModel) {
+    super.unloadInverseInternalModelFromOwn(internalModel);
+
+    const canonicalIndex = this.canonicalState.indexOf(internalModel);
+
+    if (canonicalIndex !== -1) {
+      this.canonicalState.splice(canonicalIndex, 1);
+    }
+
+    const manyArray = this._manyArray;
+
+    if (manyArray) {
+      const currentState = manyArray.currentState;
+      const idx = currentState.indexOf(internalModel);
+
+      if (idx !== -1) {
+        currentState.splice(idx, 1);
+
+        manyArray.set('length', currentState.length);
+        manyArray.notifyPropertyChange('length');
```

Comment:
if https://github.com/emberjs/data/pull/5058/files#r126607364 is addressed, most of this goes away.